src/crush/mapper.c

```diff
@@ -1081,3 +1030,1006 @@ int crush_do_rule(const struct crush_map *map,
 
 	return result_len;
 }
+
+/// invariant through crush_msr_do_rule invocation
+struct crush_msr_input {
+	const struct crush_map *map;
+	const struct crush_rule *rule;
+	
+	const unsigned result_max;
+	
+	const unsigned weight_len;
+	const __u32 *weights;
+	
+	const int map_input;
+	const struct crush_choose_arg *choose_args;
+	
+	const unsigned msr_descents;
+	const unsigned msr_collision_tries;
+};
+
+/// encapsulates work space, invariant within an EMIT block
+struct crush_msr_workspace {
+	const unsigned start_stepno;
+	const unsigned end_stepno;
+
+	const unsigned result_len;
+
+	const struct crush_work *crush_work;
+
+	// int[end_stepno - start_stepno][result_len]
+	int **step_vecs;
+};
+
+/// encapsulates output space, invariant through crush_msr_do_rule invocation
```

Comment:
invariant isn't the right word -- the output space remains the same but is mutated as items are output (out[i] and returned_so_far).  TODO